ESPS/general/src/plot_script/aplot.1
.\" Copyright (c) 1987 Entropic Speech, Inc. All rights reserved.
.\" @(#)aplot.1	3.11	4/5/90 ESI
.TH APLOT 1\-ESPS 4/5/90
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH NAME
aplot \- Plot an array of data in ASCII format.
.SH SYNOPSIS
.B aplot
[
.BI \-d
] [
.BI \-l
] [
.B \-z
] [
.BI \-T " device"
] [
.BI \-W " generic window system options"
] [
.I data_file
]
.SH DESCRIPTION
.I aplot
plots an array of ordered pairs <x, y>, stored in ASCII format.
If
.I data_file
is ``\-'' or omitted, the standard input is read.
By default,
the output is in Unix GPS (``Graphics Primitive String'') format
and is displayed on the workstation graphics display
but with the
.B \-T
option, this may be sent to the standard output,
obtained in Tektronix format rather than GPS,
or plotted on a hardcopy device.
.SH "DATA FORMAT"
.PP
The
.I data_file
is expected to contain the following parameters, each
separated by a blank.

.RI < No_of_Samples >

.RI < No_of_plots >

.RI < x_axis_min >
.RI < x_axis_max >
.RI < x_axis_step >

.RI < y_axis_min >
.RI < y_axis_max >
.RI < y_axis_step >

followed by plot data (x and y values separated by blank).
.PP
Any
.I Text
following the data (after a blank line) is reproduced as plotter output.
.PP
In the case of multiple plots, the line thickness is varied to distinguish
between different plots.
.PP
On a Sun workstation (not running X windows), 
this program must be run from within the
\fISuntools\fR window system.  The plot will appear in a new window,
which can be moved, resized, closed (reduced to an icon), and removed by
using the functions on the \fISuntools\fR menu.   To get this menu, move
the mouse cursor to the title bar of the plot window and press the right
mouse button.
.SH OPTIONS
.PP
Normally, a coninuous plot is generated.  However, the following
options can be used, in the case of single plots:
.TP
.B \-d
If this option is specified, there will be a dot for each <x, y>
pair. However, there will be no continuous plot.
.TP
.B \-l
If this option is specified, there will be a vertical line from <x,
y_axis_min> to <x, y> for each <x, y> pair; the points themselves 
will not be connected to each other.  
.TP
.B \-z
If this option is specified,
the vertical and horizontal grid lines are suppressed \- only tick
marks are made on the graph.
.TP
.BI \-T " device"
.I Device
may be
.BR gps ,
.BR mcd ,
.BR tek ,
.BR suntools ,
.BR hp ,
or
.BR hardcopy .
The default is
.BR \-Tmcd,
which displays the output on the graphics screen by piping it through
.IR mcd
or
.IR xmcd (for X window system).
The program \fImcd\fR is supplied by Masscomp for their systems and by
ESI for other systems and those running X windows.
Specifying
.B \-Tgps
sends output to the standard output in Masscomp Unix GPS format.
Specifying
.B \-Ttek
causes output to be sent to the standard output in Tektronix 4000-series
format.
If
.B \-Tsuntools
is used, then the display will be forced to use \fIsuntools\fR window
system (on Suns and Solbournes only), even if ESPS has been configured
to use the X window system.   If ESPS is built for X, then it uses X for
plotting by default.
If
.B \-Tmcd
is used on a Masscomp and ESPS has been configured to use X windows,
then the old SP-40 \fImcd\fR is used instead of X windows.
Specifying
.B \-Thardcopy
plots it on a hardcopy device, if one was set up in the installation
script.  Note that the word \fIimagen\fR was used for this option in
previous versions.   It is still accepted, but \fIhardcopy\fR is meant
to be more general.
.B \-Thp
produces output for an HP LaserJet printer.   Note that the output is
produced on standard out and it is not spooled to the printer by this
command.   Use of this option will result in about 750K bytes of output
per page.   If the link to your LaserJet is slow, it might take
several minutes to send it to the printer.    
In a later version, this will be merged with the
\fIhardcopy\fR option above for direct spooling.
.TP
.BI \-W
This option is followed by a character
string consisting of the generic window system options.   These options are
passed onto \fImcd\fR or \fIxmcd\fR.   
This option is used to pass along generic X window
arguments to \fIxmcd\fR or Suntool arguments to \fImcd\fR on Suns.  This
option might not have an effect on all systems.   For example, on a
system running X windows, to cause the plot to appear at a particular
screen location the following command could be used: 
.br
\fIaplot -W "=300x250+50+200" file\fR.
.br
See \fImcd(1\-ESPS)\fR and \fIxmcd(1\-ESPS)\fR for details.
.PP
For Sun systems, this program will plot under either X windows, or
Suntools.   By default, if the \fB-T\fR option is not used, it will 
expect X windows.   To plot under Suntools you can use the \fB-Tsuntools\fR
option.    If you always use Suntools, you can avoid using the \fB-T\fR
option by defining the environment variable \fBEPLOTDEV\fR to be
\fIsuntools\fR.  This will make the Suntools window system the default.
This variable is ignored if the machine type is not Sun4, SUN3, or
SUN386i.
.SH "ESPS PARAMETERS"
No ESPS parameter file is read.
.SH "ESPS COMMON"
The ESPS common file is not read or written.
.SH "ESPS HEADERS"
No ESPS headers are processed, since the input and output are not ESPS
files.
.SH DIAGNOSTICS
.nf
aplot: unknown option \-\fIletter\fP
Usage: aplot [\-d] [\-l] [\-z] [\-T] [input_file]
aplot: Cannot specify both \-d and \-l flags
aplot: can't open \fIfilename\fP: \fIreason\fP
aplot: EOF encountered
.SH "SEE ALSO"
plotsd(1\-ESPS), plotspec(1\-ESPS), scatplot(1\-ESPS), mcd(1G)(Masscomp),
mcd(1\-ESPS), xmcd(1\-ESPS)
.SH AUTHOR
.PP
.nf
Shankar Narayan, Entropic Speech, Inc.
Modified for \-T option by Rodney Johnson.
\fB\-z\fR option added by David Burton.
.fi


ESPS/general/src/tk_plot_tools/aplot.c
/*
 * This material contains proprietary software of Entropic Speech, Inc.
 * Any reproduction, distribution, or publication without the prior
 * written permission of Entropic Speech, Inc. is strictly prohibited.
 * Any public distribution of copies of this work authorized in writing by
 * Entropic Speech, Inc. must bear the notice
 *
 * "Copyright (c) 1997 Entropic Research Lab, Inc. All rights reserved."
 *
 *
 * Module: aplot - A program for generating x-y plots of ascii data
 *
 * Written By: S. Shankar Narayan
 * -z option added by David Burton
 * USAGE: aplot [-d] [-l] [-z] [input_file]
 * -d --- dot plot (i.e., a dot for each point)
 * -l --- line plot (i.e, vertical line for each value)
 * -z suppress grid lines
 * inp_file - input file with data and scaling information.
 *
 */
#ifndef lint
    static char *sccs_id = "@(#)aplot.c	1.3 7/21/98	ERL";
#endif

#include <stdio.h>
#include <time.h>
#include <math.h>
#include <esps/esps.h>
#include <esps/unix.h>

#define ProgName "aplot"
#define Printf (void) printf

#define SYNTAX \
USAGE("aplot [-d] [-l] [-z] [-T title_string] [input_file]");

int debug_level = 0;

main (argc, argv)
    int     argc;
    char    *argv[];
{
    int     line_plot_flag = 0,
	    dot_plot_flag = 0,
	    normal_flag = 1,
	    zflag = 0;			/* flag for suppress grid option*/
    float   xmin, xmax, xstep, ymin, ymax, ystep;
    int     color, i, j, ix, iy;
    float   x, y;
    int     newline_flag,
	    lnt;
    int	    c;
    FILE    *fd = stdin;
    extern char	*optarg;
    extern  optind;
    char    s[80];
    char    *Wflag = NULL;
    char    *filename = NULL;
    char    *title = NULL;

    while ((c = getopt (argc, argv, "dlzT:W:")) != EOF)
    {
	switch (c)
	{
	    case 'd': 
		dot_plot_flag++;
		break;
	    case 'l': 
		line_plot_flag++;
		break;
	    case 'z':
		zflag++;
		break;
	    case 'W':
		Wflag = optarg;
	        break;
	    case 'T':
		title = optarg;
	        break;
	    default: 
		SYNTAX;
	}
    }
    if (line_plot_flag && dot_plot_flag)
    {
	Fprintf(stderr, "%s: Cannot specify both -d and -l flags\n",
		ProgName);
	exit (1);
    }
    if (line_plot_flag || dot_plot_flag)
	normal_flag = 0;

    if (optind < argc - 1)
    {
	Fprintf(stderr, "%s: more than one file name.\n", ProgName);
	SYNTAX;
    }

    if (optind < argc && strcmp(argv[optind], "-") != 0) {
	TRYOPEN ("aplot", argv[optind], "r", fd);
	filename = argv[optind];
    }

    tk_init("aplot", filename, Wflag, title);
    if (fscanf (fd, "%d", &lnt) == EOF)
    {
	Fprintf(stderr, "%s: EOF encountered\n", ProgName);
	exit (1);
    }
    if (fscanf (fd, "%d %f %f %f %f %f %f",
		&color, &xmin, &xmax, &xstep, &ymin, &ymax, &ystep) == EOF)
    {
	Fprintf(stderr, "%s: EOF encountered\n", ProgName);
	exit (2);
    }


    draw_box (xmin, xmax, xstep, ymin, ymax, ystep, !zflag);

    tk_snd_plot_cmd("set_class data");
    for (j = 0; j < color; j++)
    {
	sprintf(s,"f %d", 2 * j + 1);
	tk_snd_plot_cmd(s);
	tk_snd_plot_cmd("c 3");
	for (i = 0; i < lnt; i++)
	{
	    if (fscanf (fd, "%f %f", &x, &y) != 2)
	    {
		Fprintf(stderr, "%s: EOF encountered\n", ProgName);
		exit (1);
	    }
	    iy = 3000 - 2500.0 * (y - ymin) / (ymax - ymin);
	    ix = 500 + 5000.0 * (x - xmin) / (xmax - xmin);
	    if (normal_flag)
	    {
		if (i == 0) {
		    sprintf(s,"m %d %d", iy, ix);
		    tk_snd_plot_cmd(s);
		}
		if (i != 0) {
		    sprintf(s,"d %d %d", iy, ix);
		    tk_snd_plot_cmd(s);
		}
	    }
	    if (dot_plot_flag)
	    {
		sprintf(s,"l %d %d %d %d", iy - 1, ix - 1, iy +1, ix + 1);
		tk_snd_plot_cmd(s);
		sprintf(s,"l %d %d %d %d", iy + 1, ix - 1, iy -1, ix + 1);
		tk_snd_plot_cmd(s);
	    }
	    if (line_plot_flag)
	    {
		if (ymin < 0 && ymax > 0)
		{
		    int     zero = 3000 + 2500.0 * ymin / (ymax - ymin);
		    sprintf(s,"l %d %d %d %d", zero, ix, iy, ix);
		    tk_snd_plot_cmd(s);
		}
		else
		{
		    sprintf(s,"l 3000 %d %d %d", ix, iy, ix);
		    tk_snd_plot_cmd(s);
		}
	    }
	}
    }

    tk_snd_plot_cmd("f 1");
    /* print header for the plot */
    newline_flag = 0;
    iy = 3400;
    i = getc (fd);
    while ((i = getc (fd)) > 0)
    {
	if (newline_flag == 1)
	{
	    newline_flag = 0;
	    if (i == '#')
	    {
		i = getc (fd);
/*
		sprintf(s,"f %c", i);
		tk_snd_plot_cmd(s);
		sprintf(s,"m %d 500", iy += 200);
		tk_snd_plot_cmd(s);
		tk_snd_plot_cmd("t 5 1");
*/
	    }
	    else
	    {
/*
		sprintf(s,"m %d 500", iy += 200);
		tk_snd_plot_cmd(s);
		tk_snd_plot_cmd("t 5 1");
		putchar (i);
*/
	    }
	}
/*
	else
	    putchar (i);
*/
	if (i == 13 || i == 10)
	    newline_flag = 1;

    }

    print_time (200, 4400);

    tk_start();
    exit(0);
    /*NOTREACHED*/
}

 /* Draw box */

draw_box (xmin, xmax, xstep, ymin, ymax, ystep, grid)
    float   xmin, xmax, xstep, ymin, ymax, ystep;
    int	    grid;
{
    char    label[50];
    char    s[80];
    float   tmp;
    int     ix, iy;

    tk_snd_plot_cmd("c 1");
    tk_snd_plot_cmd("s 3");
    tk_snd_plot_cmd("set_class box");

    tk_snd_plot_cmd("m 500 500");
    tk_snd_plot_cmd("d 500 5500");
    tk_snd_plot_cmd("d 3000 5500");
    tk_snd_plot_cmd("d 3000 500");
    tk_snd_plot_cmd("d  500 500");

    tk_snd_plot_cmd("c 6");
    tk_snd_plot_cmd("s 1");

    if(grid){

	/* Draw vertical grid */

        tk_snd_plot_cmd("set_class v_grid");
	for (tmp = xmin; tmp <= xmax + 0.001*xstep; tmp = tmp + xstep)
	{
	    ix = 500 + 5000 * (tmp - xmin) / (xmax - xmin);
	    tk_snd_plot_cmd_1arg("m 500 %4d", ix);
	    tk_snd_plot_cmd_1arg("d 3000 %4d", ix);

	    Sprintf(label, "%g", (fabs(tmp) < 0.001*xstep) ? 0.0 : tmp);
	    tk_snd_plot_cmd_1arg("m 3200 %4d", ix - 40*strlen(label));
			/* 1/2 width of size 5 character is 40 */
	    tk_snd_plot_cmd("t 5 1");
	    tk_snd_plot_cmd(label);
	}


	/* Draw ticks */
        tk_snd_plot_cmd("set_class ticks");
	for (tmp = xmin + xstep / 2; tmp <= xmax + 0.001*xstep; tmp = tmp + xstep)
	{
	    ix = 500 + 5000 * (tmp - xmin) / (xmax - xmin);
	    tk_snd_plot_cmd_1arg("m 475 %4d", ix);
	    tk_snd_plot_cmd_1arg("d 525 %4d", ix);
	    tk_snd_plot_cmd_1arg("m 2975 %4d", ix);
	    tk_snd_plot_cmd_1arg("d 3025 %4d", ix);
	}

	/* Draw horizontal grid */
        tk_snd_plot_cmd("set_class h_grid");
	for (tmp = ymin; tmp <= ymax + 0.001*ystep; tmp = tmp + ystep)
	{
	    iy = 3000 - 2500 * (tmp - ymin) / (ymax - ymin);
	    tk_snd_plot_cmd_1arg("m %4d 5500", iy);
	    tk_snd_plot_cmd_1arg("d %4d 500", iy);
	    Sprintf(label, "%g", (fabs(tmp) < 0.001*ystep) ? 0.0 : tmp);
	    sprintf(s,"m  %4d %d", iy + 50, 400 - 80*strlen(label));
			/* 1/2 height of size 5 digit is 50 */
	    tk_snd_plot_cmd(s);
	    tk_snd_plot_cmd("t 5 1");
	    tk_snd_plot_cmd(label);
	}
    }
    else
    {/* don't draw the grid - just ticks */
	/* Draw x axis ticks */

        tk_snd_plot_cmd("set_class ticks");
	for (tmp = xmin; tmp <= xmax + 0.001*xstep; tmp = tmp + xstep)
	{
	    ix = 500 + 5000 * (tmp - xmin) / (xmax - xmin);
	    tk_snd_plot_cmd_1arg("m 475 %4d", ix);
	    tk_snd_plot_cmd_1arg("d 525 %4d", ix);
	    tk_snd_plot_cmd_1arg("m 2975 %4d", ix);
	    tk_snd_plot_cmd_1arg("d 3025 %4d", ix);

	    Sprintf(label, "%g", (fabs(tmp) < 0.001*xstep) ? 0.0 : tmp);
	    tk_snd_plot_cmd_1arg("m 3200 %4d", ix - 40*strlen(label));
			/* 1/2 width of size 5 character is 40 */
	    tk_snd_plot_cmd("t 5 1");
	    tk_snd_plot_cmd(label);
	}

	/* Draw the y axis ticks*/
	for (tmp = ymin; tmp <= ymax + 0.001*ystep; tmp = tmp + ystep)
	{
	    iy = 3000 - 2500 * (tmp - ymin) / (ymax - ymin);
	    tk_snd_plot_cmd_1arg("m %4d 525", iy);
	    tk_snd_plot_cmd_1arg("d %4d 475", iy);
	    Sprintf(label, "%g", (fabs(tmp) < 0.001*ystep) ? 0.0 : tmp);
	    sprintf(s,"m  %4d %d", iy + 50, 400 - 80*strlen(label));
			/* 1/2 height of size 5 digit is 50 */
	    tk_snd_plot_cmd(s);
	    tk_snd_plot_cmd("t 5 1");
	    tk_snd_plot_cmd(label);
	}
    }
    tk_snd_plot_cmd("c 4");
}
